src/librustc/ty/instance.rs

```diff
@@ -279,7 +278,7 @@ impl<'tcx> Instance<'tcx> {
         def_id: DefId,
         substs: SubstsRef<'tcx>,
     ) -> Option<Instance<'tcx>> {
-        (*RESOLVE_INSTANCE)(tcx, param_env, def_id, substs)
+        tcx.resolve_instance((tcx.erase_regions(&param_env), def_id, tcx.erase_regions(&substs)))
```

Comment:
Hmm, I am concerned about the use of `erase_regions` here. It feels potentially surprising to me, particularly since the returned `Instance` type can embed types/regions from the input (which would now be erased). Do you know what the impact of this PR is without this call?

Note that a lot of the folks invoking this (e.g., during trans and other passes) will already have erased regions.

Reply:
If we don't erase regions, we'd need to make this a canonical query, as [this ends up getting invoked with region variables](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/22006/logs/68). I don't see a straightforward way of benchmarking the difference, since we'd be adding the cost of canonicalization.

Do any of the callers care about the regions in the returned `Instance`? It looks like `Instance::resolve` is called by CTFE, codegen, and some symbol-related code, none of which (I think) should care about the regions in the returned `Instance`.

Reply:
Actually, it turns out that `Instance::resolve` already ends up erasing regions. We either end up calling `tcx.codegen_fulfill_obligation` (which erases regions):

https://github.com/rust-lang/rust/blob/dd67187965e136bff1ed05e035293441c60f0790/src/librustc_ty/instance.rs#L73

or we call `tcx.subst_and_normalize_erasing_regions`, which also erases:

https://github.com/rust-lang/rust/blob/dd67187965e136bff1ed05e035293441c60f0790/src/librustc_ty/instance.rs#L22

I think erasing regions before we invoke `tcx.resolve_instance` won't actually affect the output - we just end up erasing regions slightly earlier than we otherwise would.

Reply:
MIR optimizations may use `Instance::resolve`, but yeah, no result of `Instance::resolve` should be flowing into borrow-checking.